### ISSUE TITLE
Some Input bugs (next)

### ISSUE THREAD

**Issue by mjashanks** [Budibase/budibase#1482]
Here are a few issues that I encountered on the `next` branch, around Inputs.

**Input values are `null` when empty.** 
When setting up a query for a rest API (external datasource) i didn't set the "Query String" field. When clicking save, I got a weird error. It turned out to be because the the "Query String" field was null. Entering a space into the field allowed me to save. 

This might be a matter of `inputValue ?? ""` on that particular area, but I am flagging as it looks like a regression due to new input behaviour.

**Underlying value only changes on blur**
When creating a new app, the "Next" button does not enable until I click/tab out. Same with creating a new datasource.

It does not behave in this way on master

**Toggles look weird on new table dialog (I'm using Ubuntu & Firefox)**
![image](https://user-images.githubusercontent.com/3524181/117954829-519c2100-b30f-11eb-800e-6458ebe1d5a4.png)

**Comment by aptkingston** [comment]
Alright so there are a couple of things to discuss here. At the minute none of these are bugs in that they're all by design (other than the weird checkbox thing).

The initial values of all type of all form fields is `null`, except toggle/checkbox which are explicitly boolean, for consistency. This is similar to how Svelte handles things, where if you bind an input value to an input, the value will be undefined until you type in the field. If you type in a field and clear it, the value will be an empty string `""`. It's only `null` before it has been touched. I wouldn't be totally against changing text inputs to default to an empty string, but it makes it slightly weird for inputs of type `number` as they normally emit real numeric values. It becomes messy for things like selects too - should they default to empty string? In my opinion they should definitely be null. I am not a fan of empty strings in general, but I'm happy to chat about this and see what makes the most sense for us. IMO this is where JS is a mess, when it comes to `undefined` vs `null` vs `""` vs `NaN` etc.

The change on blur is something I'm aware of and I'm going to address. The reason it needed to be on blur (rather than on change) was because of the way we auto-save component settings. Before spectrum we used different components for component settings (which updated on blur) than for other places like that create table modal (which updated on change). I picked the less aggressive of these for the unified components to avoid the guaranteed 409 document update conflicts and just general load that would happen if we fired off change requests on every keystroke in component settings. I realise the update-on-blur approach is frustrating for forms with validation (like the create table modal), so I'll look at adding props to control when the value is updated, either on change or on blur.

I tested the checkbox issue there and I can only reproduce it on Firefox on Linux as well. That must be actually a Spectrum bug. I'll look at adding some custom CSS to fix it.

**Comment by mjashanks** [comment]
Yea - that all makes sense. 

RE: input defaults to null. I agree with your points - I am just flagging that this caused an actual bug for me. This tells me that there could be more bugs caused by the same thing - i.e. it's a breaking change in the API of the `input`.

**Comment by aptkingston** [comment]
Good catch then - that's not something I tested. I did go through and test (what I thought) was all of the builder because there were lots of consequences like that with the Spectrum update, but I definitely didn't test each individual data source type so we need to fix that.

**Comment by aptkingston** [comment]
Fixed the other part of this via one of my recent PR's. Most text inputs now update immediately, except for the component settings which are on blur. The checkbox issue should also be fixed, and should not be visible any more.
